src/linker/Linker/DynamicDependency.cs

```diff
@@ -0,0 +1,140 @@
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Mono.Cecil;
+
+#nullable enable
+
+namespace Mono.Linker
+{
+	/// Tracks dependencies created via DynamicDependencyAttribute in the linker.
+	/// This is almost identical to DynamicDependencyAttribute, but it holds a
+	/// TypeReference instead of a Type, and it has a reference to the original
+	/// CustomAttribute for dependency tracing. It is also a place for helper
+	/// methods related to the attribute.
+	internal class DynamicDependency : Attribute
```

Comment:
Why does it derive from Attribute?

Reply:
Because LinkerAttributesInformation enforces "Attribute" base class. Sven actually originally changed that to be "object" based, but in the end we didn't like it. I still think it's better for these to be attribute based - it makes it clearer what the class represents.
But if others don't like it I'll change the cache to be object based.

Reply:
It's weird to see Attribute like without AttributeUsage. What does Attribute give us here?

Reply:
It doesn't give us anything functionality - that said we could use it - we could enforce `AttributeUsage` in the common code. But it's not very useful as it's highly unlikely we'll get an attribute on the wrong item (since C# would not allow it anyway).
I'm fine if we agree to change it to object based approach.

Reply:
I plan to leave it as-is unless you would like me to change the cache to be object-based instead @marek-safar.